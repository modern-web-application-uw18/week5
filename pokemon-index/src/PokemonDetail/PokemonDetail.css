._PokemonDetail {
    font-size: 15px;
    font-family: Monaco;
    display: flex;
    flex-direction: column;
    align-items: center;
}

.CardContainer {
    border: 1px solid black;
    width: 350px;
    height: 600px;
    margin: 0 auto;
    margin-top: 56px;
    border-radius: 25px;
    box-sizing: border-box;
    box-shadow: -8px 9px 16px -3px gray;
    background: rgb(252, 252, 74);
}

.CardBackground {
    height: 550px;
    margin: 20px 20px 0 20px;
    background: rgb(247, 182, 61);
    border: solid orange;
    border-top-left-radius: 6px;
    border-top-right-radius: 6px;
    border-bottom-left-radius: 6px;
    border-bottom-right-radius: 6px;
}

.CardFrame {
    font-size: 15px;
    font-family: Monaco;
    display: flex;
    flex-direction: column;
}

.ImageBackground {
    height: 140px;
    margin: 0px 20px 20px 20px;
    width: 260px;
    background: url(./background.jpeg);
    text-align: center;
}

.PokemonDetailHeader {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    background-color: orange;
}

.PokemonDetailName {
    margin-left: 15px;
    font-size: 20px;
    font-weight: bold;
    flex-grow: 2;
}

.PokemonTypeSprite {
    width: 30px;
    margin-right: 10px;
}

.PokemonDetailHP {
    margin-right: 3px;
}

.PokemonDetailBody {
    margin-top: 15px;
    display: flex;
    flex-direction: column;
    align-items: center;
    
}

