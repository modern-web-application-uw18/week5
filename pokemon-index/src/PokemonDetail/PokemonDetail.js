import React, { Component } from 'react';
import './PokemonDetail.css';
import PropTypes from 'prop-types';

class PokemonDetail extends Component {
    constructor(props) {
        super(props);
        this.state = {
            pokemonInfo: {}
        }
    }

    typeToImageLocation = {
        fire: "/fire.png",
        grass: "/grass.png",
        bug: "/grass.png",
        poison: "/grass.png",
        water: "/water.png",
        ice: "/water.png",
        electric: "/lightning.png",
        fighting: "/fighting.png",
        rock: "/fighting.png",
        ground: "/fighting.png",
        psychic: "/psychic.png",
        ghost: "/psychic.png",
        normal: "/colorless.png",
        flying: "/colorless.png",
        dark: "/darkness.png",
        steel: "/metal.png",
        dragon: "/dragon.png",
        fairy: "/fairy.png"
    };

    componentDidMount() {
        //match is provided by Route (injected by the Route component)
        const pokemonId = this.props.match.params.pokemonId;
        fetch('https://pokeapi.co/api/v2/pokemon/'+pokemonId+'/')
            .then(response => response.json())
            .then(data => {
                this.setState((prevState, props) => {
                    return {
                        pokemonInfo: data
                    }
                })
            })
            .catch(error => console.log(error)
        );
    }
  
    render() {
        let obj = this.state.pokemonInfo; 
        let imageURL = "";
        let typeArray = [];
        let mainType = ""; //primary type
        let mainTypeSpriteSrc = "";
        let statsSummary = [];
        let baseExp = "";
        let hp = 0;
        let statsObj = {};
        // build imageURL only when state is populated
        if (Object.keys(obj).length !== 0 && obj.constructor === Object) {
            imageURL = this.state.pokemonInfo.sprites.front_shiny;
            typeArray = this.state.pokemonInfo.types.map((idx,pokemonType) => {
                if (this.state.pokemonInfo.types[pokemonType].slot === 1) {
                    mainType = this.state.pokemonInfo.types[pokemonType].type.name;
                    mainTypeSpriteSrc = `${this.typeToImageLocation[mainType]}`;            }
                return (this.state.pokemonInfo.types[pokemonType].type.name);
            });
            typeArray = "(" + typeArray.join(', ') + ")";
            statsSummary = this.state.pokemonInfo.stats.filter((stat) => {
                if (stat.stat.name === 'hp') {
                    hp = stat.base_stat;
                    return false;  //hp is already displayed in the header
                }
                return true;
            }).map((statIdx,stat) => {
                statsObj[this.state.pokemonInfo.stats[stat].stat.name] = this.state.pokemonInfo.stats[stat].base_stat;
                return (
                    <p key={stat+statIdx}>
                        {this.state.pokemonInfo.stats[stat].stat.name}:  {this.state.pokemonInfo.stats[stat].base_stat}
                    </p>
                );
            });
            baseExp = "base exp: " + this.state.pokemonInfo.base_experience;
        }

        return (
            <div className="CardContainer">
                <div className="CardBackground">
                    <div className="CardFrame">
                        <div className="PokemonDetailHeader">
                            <p className="PokemonDetailName">{this.state.pokemonInfo.name}</p>
                            <img className = "PokemonTypeSprite" src={mainTypeSpriteSrc} alt="Pokemon Type Sprite"/>
                            <p className="PokemonDetailHP">HP {hp}</p>
                        </div>
                        <div className="PokemonDetailBody">
                            <div className="ImageBackground">
                                <p><img src={imageURL} alt="Pokemon sprite"/></p>
                            </div>
                            <div className="PokemonDetailStats">
                                <p className="PokemonDetailType">{typeArray}</p>
                                <p className="PokemonDetailBaseExp">{baseExp}</p>
                                {statsSummary}
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        )
    }
  }

  PokemonDetail.propTypes = {
    match: PropTypes.array
}

  export default PokemonDetail;