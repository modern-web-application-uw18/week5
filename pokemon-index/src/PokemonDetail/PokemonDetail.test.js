import React from 'react';
import ReactDOM from 'react-dom';
import PokemonDetail from './PokemonDetail.js';
import { MemoryRouter as Router, Route } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';

it('renders without crashing', () => {
    const div = document.createElement('div');
    const component = (
        <Router>
            <Route exact path="/pokemon/1" component={PokemonDetail} />
        </Router>
    );
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
});

it('should render a Pokemon detail page', () => {
    const div = document.createElement('div');
    const component = TestRenderer.create(
        <Router>
            <Route exact path="/pokemon/1" component={PokemonDetail} />
        </Router>
    );
    expect(component.toJSON()).toMatchSnapshot();
});