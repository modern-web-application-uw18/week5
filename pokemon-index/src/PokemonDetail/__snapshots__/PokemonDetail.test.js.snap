// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a Pokemon detail page 1`] = `null`;
